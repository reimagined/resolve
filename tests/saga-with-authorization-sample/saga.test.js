import interopRequireDefault from '@babel/runtime/helpers/interopRequireDefault'
import givenEvents from 'resolve-testing-tools'

import config from './config'

describe('Saga', () => {
  const {
    name: sagaName,
    source: sourceModule,
    connectorName
  } = config.sagas.find(({ name }) => name === 'ProcessKiller')
  const {
    module: connectorModule,
    options: connectorOptions
  } = config.readModelConnectors[connectorName]

  const createConnector = interopRequireDefault(require(connectorModule))
    .default
  const source = interopRequireDefault(require(`./${sourceModule}`)).default

  const { commands: commandsModule } = config.aggregates.find(
    ({ name }) => name === 'Process'
  )

  const commands = interopRequireDefault(require(`./${commandsModule}`)).default

  let commandTimestamp = 0
  const executeCommand = ({ aggregateId, type, jwtToken, ...command }) => {
    const event = {
      ...commands[type]({}, command, jwtToken),
      aggregateId,
      timestamp: commandTimestamp++
    }
    return event
  }

  let sagaWithAdapter = null
  let adapter = null

  beforeEach(async () => {
    adapter = createConnector(connectorOptions)
    try {
      const connection = await adapter.connect(sagaName)
      await adapter.drop(null, sagaName)
      await adapter.disconnect(connection, sagaName)
    } catch (e) {}

    sagaWithAdapter = {
      handlers: source.handlers,
      sideEffects: source.sideEffects,
      adapter
    }
  })

  afterEach(async () => {
    try {
      const connection = await adapter.connect(sagaName)
      await adapter.drop(null, sagaName)
      await adapter.disconnect(connection, sagaName)
    } catch (e) {}

    adapter = null
    sagaWithAdapter = null
  })

  test('success registration', async () => {
    const result = await givenEvents([
      executeCommand({
        aggregateId: 'id1',
        aggregateName: 'Process',
        type: 'createProcess'
      }),
      executeCommand({
        aggregateId: 'id2',
        aggregateName: 'Process',
        type: 'createProcess'
      }),
      executeCommand({
        aggregateId: 'id3',
        aggregateName: 'Process',
        type: 'createProcess'
      }),
      executeCommand({
        aggregateId: 'root',
        aggregateName: 'Process',
        type: 'killAllProcesses'
      })
    ]).saga(sagaWithAdapter)

    expect(result).toMatchSnapshot()
  })
})
