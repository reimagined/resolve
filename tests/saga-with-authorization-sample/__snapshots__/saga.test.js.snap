// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Saga success registration 1`] = `
Object {
  "commands": Array [
    Array [
      Object {
        "aggregateId": "id1",
        "aggregateName": "Process",
        "jwtToken": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJwZXJtaXNzaW9ucyI6eyJwcm9jZXNzZXMiOnsia2lsbCI6dHJ1ZX19fQ.S6z8Dml89H9DsrFH-6H37EmDn0e7r6caPUQHr2b3vow",
        "type": "PROCESS_KILLED",
      },
    ],
    Array [
      Object {
        "aggregateId": "id2",
        "aggregateName": "Process",
        "jwtToken": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJwZXJtaXNzaW9ucyI6eyJwcm9jZXNzZXMiOnsia2lsbCI6dHJ1ZX19fQ.S6z8Dml89H9DsrFH-6H37EmDn0e7r6caPUQHr2b3vow",
        "type": "PROCESS_KILLED",
      },
    ],
    Array [
      Object {
        "aggregateId": "id3",
        "aggregateName": "Process",
        "jwtToken": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJwZXJtaXNzaW9ucyI6eyJwcm9jZXNzZXMiOnsia2lsbCI6dHJ1ZX19fQ.S6z8Dml89H9DsrFH-6H37EmDn0e7r6caPUQHr2b3vow",
        "type": "PROCESS_KILLED",
      },
    ],
  ],
  "queries": Array [],
  "scheduleCommands": Array [],
  "sideEffects": Array [],
}
`;
